webapp/src/components/Analysis/PerturbationTestingPreview.test.tsx

```diff
@@ -0,0 +1,243 @@
+    await waitFor(() => {
+      //Verifying if the train toggle button is disabled if not available in datasplits
+      expect(screen.getByRole("button", { pressed: false })).toHaveClass(
+        "Mui-disabled"
+      );
+      // By default, evaluation toggle is selected on page load.
+      expect(screen.getByRole("button", { pressed: true })).not.toHaveClass(
+        "Mui-disabled"
+      );
+    });
+  });
+
+  it("should display the data in the toggle with expected styling (color, font), format", async () => {
+    renderPerturbationTestingPreview({ train: true, eval: true });
+    await waitFor(() => {
+      // Verrify if the actual response data from API is visibly displayed on the toggle
+      expect(screen.getByText(/14.0%/i)).toBeVisible();
+      expect(screen.getByText("Failure rate - Evaluation Set")).toBeVisible();
+      expect(screen.getByText(/16.0%/i)).toBeVisible();
+      expect(screen.getByText("Failure rate - Training Set")).toBeVisible();
+
+      // styling should be applied for the selected toggle as expected
+      expect(screen.getByText(/14.0%/)).toHaveStyle(
+        "font-size: 3.75rem; color: rgb(237, 108, 2)"
+      );
+
+      // selected toggle should be highlighted with some bgcolor
+      expect(screen.getByRole("button", { pressed: true })).toHaveStyle(
+        "color: rgb(156, 39, 176)"
+      );
+
+      // verify with some random unexpected values to not have been displayed.
+      expect(screen.queryByText(/17.0%/i)).toBeNull();
+      expect(screen.queryByText("some random text")).toBeNull();
+      expect(screen.queryByText(/26.0%/i)).toBeNull();
+    });
+  });
+
+  it("should not display value if a datasplit is not available", async () => {
+    renderPerturbationTestingPreview({ train: false, eval: true });
+    await waitFor(() => expect(screen.getByText("--%")).toBeVisible());
+  });
+
+  it("should verify the perturbation preview table displayed with expected columns, rows, formatted contents, sort model", async () => {
+    renderPerturbationTestingPreview({ train: true, eval: true });
+    await waitFor(() => {
+      const table_columns = [
+        "Test Family",
+        "Test Name",
+        "Modif. Type",
+        "Test Description",
+        "FR on Evaluation Set",
+      ];
+      const columheaders = screen.getAllByRole("columnheader");
+
+      // verify if all the required columns are displayed
+      columheaders.forEach((item, index) => {
+        expect(item.textContent).toEqual(table_columns[index]);
+      });
+    });
+  });
+
+  it("should be having a default sort as desc for colum header 'FR on Evaluation/Training set'", async () => {
+    renderPerturbationTestingPreview({ train: true, eval: true });
+    const columheaders = await screen.findAllByRole("columnheader");
+    const fr_column = columheaders.find(
+      (column) => column.ariaColIndex === "5"
+    );
+    fr_column && waitFor(() => expect(fr_column.ariaSort).toBe("descending"));
```

Comment:
With `fr_column && `, if we don't find a column with `ariaColIndex === "5"`, we silently don't test anything. I guess that's not desired. Also, I'm sure we can get the right column header directly, without having to loop. Can you try that?

Reply:
Actually, we pass the API data in the beginning so it should have columns and rows but null check was actually required by the syntax rule here. But I dont like to do that as well. if we have a better approach to deal this, it would be great to know.

And I tried fetching like giving columnheader[index] but test could not find the expected index/header. so had to do loop and find the index.

Reply:
Well the test passes because it doesn't test anything. Two reasons:
* `fr_column` is `undefined`
* You did not `await` the promise returned by `waitFor()`, so `() => expect(fr_column.ariaSort).toBe("descending")` would never be executed anyway.

So typing gets in the way here with `fr_column` being possibly `undefined`? You can assert that it is not `undefined` with the `!` post-fix "Non-null assertion operator".
```suggestion
    const columheaders = await screen.findAllByRole("columnheader");
    const fr_column = columheaders.find(
      (column) => column.ariaColIndex === "5"
    );
    await waitFor(() => expect(fr_column!.ariaSort).toBe("descending"));
```

You could also add a `expect(fr_column).toBeInTheDocument();` to make the error message clearer. But again, I'm sure we can get the right column header directly, without having to loop.

Reply:
By the way, the trick of trying the test with an unexpected value would have worked to make sure it tests something. For example, you could have changed `"descending"` to `"potato"`, and the test would have still passed, indicating it is not doing its job.
```suggestion
    const columheaders = await screen.findAllByRole("columnheader");
    const fr_column = columheaders.find(
      (column) => column.ariaColIndex === "5"
    );
    fr_column && waitFor(() => expect(fr_column.ariaSort).toBe("potato"));
```